src/gt4py/next/program_processors/runners/dace/transformations/gpu_utils.py

```diff
@@ -651,27 +656,61 @@ def apply(
         graph: Union[dace.SDFGState, dace.SDFG],
         sdfg: dace.SDFG,
     ) -> None:
-        """Modify the map as requested."""
+        """Modify the map as requested.
+
+        The block size is set according to the following rules:
+        - Determine the number of map parameters that have range > 1.
+        - If there is only one such parameter then handle the map as 1D map,
+          else if there are more than 2 parameters with range > 1 then handle it as ordinary map.
+        - For 1D maps set the block size such that only the dimension with the largest
+          range is set to the specified block size, all other dimensions are set to 1.
+        - If launch_bounds_1d is not set then set the launch_bounds of `scan` maps to 512 to limit their register usage.
+        """
         gpu_map: dace_nodes.Map = self.map_entry.map
         map_size = gpu_map.range.size()
         dims_to_inspect = len(map_size)
         num_map_params = 0
         # Order of map parameters is from outer to inner, i.e. z,y,x
-        for i, axis_size in enumerate(map_size):
-            if i < len(map_size) - 1 and map_size[i + 1] == 1:
-                assert axis_size <= 1, (
-                    "GPU thread block size setting currently does not support maps where non-leading "
-                    "dimensions have size greater than one if the next dimension has size one."
-                )
+        for axis_size in map_size:
             if axis_size != 1:
-                num_map_params += 1  # Handle 2D maps where one dimension has range 1 as 1D map
+                num_map_params += 1  # Count how many parameters have range > 1
+        # If the number of map parameters that have range > 1 is 1 then handle the map as 1D map else handle it as full dimensional map
+        if num_map_params > 1 and dims_to_inspect > num_map_params:
```

Comment:
For your information, the pythonic way of doing it would be:
```suggestion
        if dims_to_inspect > num_map_params > 1:
```
However, I am not a fan of this chaining either, so you can keep it the way it is.